solidity/contracts/deposit/DepositUtils.sol

```diff
@@ -447,24 +447,58 @@ library DepositUtils {
+        address _redeemer
+    ) internal view returns(uint256 owedToDeposit, uint256 owedToRedeemer) {
+        uint256 baseRedemptionRequirement = computeRedemptionFeeAdjustment(_d, _redeemer);
+        uint256 currentBalance = _d.tbtcToken.balanceOf(address(this));
+        if(currentBalance > baseRedemptionRequirement){
+            owedToRedeemer = currentBalance.sub(baseRedemptionRequirement);
+            return(0, owedToRedeemer);
+        }
+        owedToDeposit = baseRedemptionRequirement.sub(currentBalance);
+        return(owedToDeposit, 0);
     }
 
     /// @notice             Get TBTC amount required for redemption by a specified _redeemer.
     /// @dev                Will revert if redemption is not possible by msg.sender.
     /// @param _redeemer    The deposit redeemer.
-    /// @return             The amount in TBTC needed to redeem the deposit.
-    function getRedemptionTbtcRequirement(DepositUtils.Deposit storage _d, address _redeemer) internal view returns(uint256) {
+    /// @return             Tuple, first value represents the amount in TBTC needed to redeem
+    ///                     the deposit. The second value represents the amount in TBTC the deposit
+    ///                     owes to the redeemer in case of overpayment by an external source.
+    function getRedemptionTbtcRequirement(
+        DepositUtils.Deposit storage _d,
+        address _redeemer
+    ) internal view returns(uint256 owedToDeposit, uint256 owedToRedeemer, uint256 owedToTdtHolder) {
         bool redeemerHoldsTdt = depositOwner(_d) == _redeemer;
         bool inCourtesy = _d.inCourtesyCall();
         bool atTerm = remainingTerm(_d) == 0;
 
         require(redeemerHoldsTdt || inCourtesy || atTerm, "Only TDT holder can redeem unless deposit is at-term or in COURTESY_CALL");
 
         uint256 mainCharge = computeBaseRedemptionCharge(_d, redeemerHoldsTdt);
-        return mainCharge.add(computeRedemptionFeeAdjustment(_d, _redeemer));
+        uint256 baseRedemptionRequirement = mainCharge.add(computeRedemptionFeeAdjustment(_d, _redeemer));
+        uint256 currentBalance = _d.tbtcToken.balanceOf(address(this));
+        if(currentBalance == 0){
+            if(!redeemerHoldsTdt){
+                owedToTdtHolder = lotSizeTbtc(_d).sub(signerFee(_d));
+            }
+            return(baseRedemptionRequirement, 0, owedToTdtHolder);
+        }
+        if(!redeemerHoldsTdt){
+            if(currentBalance >= signerFee(_d)){
+                owedToTdtHolder = lotSizeTbtc(_d);
+            }else{
+                owedToTdtHolder = lotSizeTbtc(_d).sub(currentBalance);
+            }
+        }
+        if(currentBalance > baseRedemptionRequirement){
+            owedToRedeemer = currentBalance.sub(baseRedemptionRequirement);
+            return(0, owedToRedeemer, owedToTdtHolder);
+        }
+        owedToDeposit = baseRedemptionRequirement.sub(currentBalance);
+        return(owedToDeposit, 0, owedToTdtHolder);
```

Comment:
Ok, so I think there's value to stacking these but I think the approach is a bit jumbled right now. I suggest we compute everything with respect to the initial escrow amount.

In particular, we first look up how much *must be in escrow*.

Then, we compute:

- the amount owed to the deposit as *escrow requirement - current balance*, clamped to 0.
- the amount owed to the redeemer as *current balance - escrow requirement*, clamped to 0.
- the amount owed to the TDT holder as *escrow requirement - (signer fee - fee rebate)*, with fee rebate only being included if `preTerm && ! _redeemerHoldsFrt`.
- the amount owed to the FRT holder as *fee rebate*, with a max of current balance. We don't currently return this, but that seems like an oversight?

I would introduce a helper here:

```
function subtractWithFloor(uint256 left, uint256 right, uint256 min) returns (uint256) {
    if (left > right) {
        return left.sub(right);
    } else {
        return min;
    }
}
```

To make this read more clearly.

Thoughts? This removes almost all conditionals from the body of the function, except for the one around fee rebate.

Reply:
I'm not sure I totally follow

Reply:
```
        uint256 mainCharge = computeBaseRedemptionCharge(_d, redeemerHoldsTdt);
        uint256 baseRedemptionRequirement = mainCharge.add(computeRedemptionFeeAdjustment(_d, _redeemer));
        uint256 currentBalance = _d.tbtcToken.balanceOf(address(this));

        owedToDeposit = subtractWithFloor(baseRedemptionRequirement, currentBalance, 0);
        owedToRedeemer = subtractWithFloor(currentBalance, baseRedemptionRequirement, 0);
        if(!redeemerHoldsTdt){
            owedToTdtHolder = lotSize.sub(subtractWithFloor(signerFee, currentBalance, 0));
        }
        return(owedToDeposit, owedToRedeemer, owedToTdtHolder);
```
is a simplified state using the helper that works as well

Reply:
> the amount owed to the FRT holder as fee rebate, with a max of current balance. We don't currently return this, but that seems like an oversight?

Not fully clear on what the FRT issue is. All we want to do is escrow what is needed. The fee is distributed (if it is escrowed) later on in the redemption flow. What is missing from the FRT handling logic?

Reply:
Do we pay all of these pieces out at the beginning of the redemption flow? I'm trying to see where the logical separation is between the owed amounts we return here and the owed amount for the FRT. If we can't compute that because it's managed later in the flow, we have an opportunity to note that in the function docs.

Reply:
Also---the simplified logic with the helper looks much more readable to me, other than a few nits I have with a couple of the variable names---wdyt?

Reply:
during `provideRedemptionProof`, we'll `distributeSignerFee`, then `distributeFeeRebate`. These are callen in succession and they read from the Deposit balance. `distributeSignerFee` should always do something since we should always have `signerFee` escrowed at this time, but `distributeFeeRebate` is not guaranteed to do something. These both depend on the return value from `getRedemptionTbtcRequirement`. The owed amounts should include the FRT. `computeRedemptionFeeEscrow` returns `signerFee` amount as base return value, or  `2*signerFee` if fee rebate is owed. 
`performRedemptionTBTCTransfers` enforces that these values are escrowed in the Deposit, and `provideRedemptionProof` will finally handle sending the `signerFee` and `FeeRebate`

Reply:
It's worth explicitly writing these invariants down somewhere, probably in the docs for `performRedemptionTBTCTransfers` (something along the lines of “once these transfers complete, the deposit balance should be sufficient to pay out signer fees plus any fee rebate owed”).